api/src/main/java/bio/terra/tanagra/service/databaseaccess/bigquery/BigQueryExecutor.java

```diff
@@ -0,0 +1,69 @@
+package bio.terra.tanagra.service.databaseaccess.bigquery;
+
+import bio.terra.common.exception.InternalServerErrorException;
+import bio.terra.tanagra.service.databaseaccess.QueryExecutor;
+import bio.terra.tanagra.service.databaseaccess.QueryRequest;
+import bio.terra.tanagra.service.databaseaccess.QueryResult;
+import bio.terra.tanagra.service.databaseaccess.RowResult;
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQuery.QueryResultsOption;
+import com.google.cloud.bigquery.BigQueryException;
+import com.google.cloud.bigquery.FieldValueList;
+import com.google.cloud.bigquery.Job;
+import com.google.cloud.bigquery.JobInfo;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableResult;
+import com.google.common.collect.Iterables;
+import java.time.Duration;
+import javax.servlet.http.HttpServletResponse;
+
+/** A {@link QueryExecutor} for Google BigQuery. */
+public class BigQueryExecutor implements QueryExecutor {
+  private static final Duration MAX_QUERY_WAIT_TIME = Duration.ofSeconds(60);
+
+  /** The BigQuery client to use for executing queries. */
+  private final BigQuery bigQuery;
+
+  public BigQueryExecutor(BigQuery bigQuery) {
+    this.bigQuery = bigQuery;
+  }
+
+  @Override
+  public QueryResult execute(QueryRequest queryRequest) {
+    QueryJobConfiguration queryJobConfiguration =
+        QueryJobConfiguration.newBuilder(queryRequest.sql()).build();
+    // TODO add pagination. Right now we always only return the first page of results.
+    TableResult tableResult = getResults(queryJobConfiguration);
+
+    Iterable<RowResult> rowResults =
+        Iterables.transform(
+            tableResult.getValues(),
+            (FieldValueList fieldValueList) ->
+                new BigQueryRowResult(fieldValueList, queryRequest.columnHeaderSchema()));
```

Comment:
Seems like the header schema is mostly passed through from the reques t- at what point would we expect to see an error from a mismatch between the expected header schema and the actual row results?

Reply:
- yes, indirectly from the attributes and column mapping. (and also the result columns could be from the selection, e.g. count(x) -> int)
- We get the error when the value is retrieved. We could do it when we do the wrapping of FieldValue -> CellValue, which is a little earlier but not much better. I don't know we can figure it out earlier than that.

The assert I was referring to is only for the number of columns & number of FieldValues, in the BigQueryRowResult constructor (sorry, misleading). https://github.com/DataBiosphere/tanagra/blob/1bcb8c6d91623e8b652fb3a44494cbae547d00c6/api/src/main/java/bio/terra/tanagra/service/databaseaccess/bigquery/BigQueryRowResult.java#L14-L19